### PR TITLE
Fix #1307: Upgrade adblock stats library to use adblock-rust.

### LINKED ISSUE #1307: Move from old adblock stats lib to adblock-rust

### PR DESCRIPTION
**Pull request by iccub**
For now there is no automation when it comes to updating this library, it needs to be done manually.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [x] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)
- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New files have MPL-2.0 license header.

## Test Plan:

Browse through popular websites and see if adblock stats are updating and the app is not crashing.
Play few youtube/twitch videos as well

## Reviewer Checklist:

- [x] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [x] Issues are assigned to at least one epic.
- [x] Issues include necessary QA labels:
  - [x] `QA/(Yes|No)`
  - [x] `release-notes/(include|exclude)`
  - [x] `bug` / `enhancement`
- [x] Necessary security reviews have taken place.
- [x] Adequate test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable)

### CODE REVIEW

File: Client/WebFilters/AdblockResourcesMappings.swift [hunk 1]
```diff
@@ -5,7 +5,14 @@
 import Foundation
 
 struct AdblockResourcesMappings {
-    static let generalAdblockName = "latest"
+    static func generalAdblockName(for fileType: FileType) -> String? {
+        switch fileType {
+        case .dat: return "rs-ABPFilterParserData"
```
Comment (iccub) [inline Client/WebFilters/AdblockResourcesMappings.swift:10]: Some logic had to be added to handle differences between ios-json and universal-dat files, this includes different endpoints and file names

File: Client/WebFilters/ShieldStats/Adblock/Resources/ablock_rust_lib.h [hunk 2]
```diff
@@ -0,0 +1,96 @@
+#ifndef ADBLOCK_RUST_FFI_H
```
Comment (iccub) [inline Client/WebFilters/ShieldStats/Adblock/Resources/ablock_rust_lib.h:1]: This file was generated, see https://github.com/brave/adblock-rust-ffi#regenerating-the-c-header

File: Client/WebFilters/AdblockResourcesMappings.swift [hunk 3]
```diff
@@ -14,37 +21,45 @@ enum ResourceLocale: String {
     case ar, bg, zh, cs, de, da, et, fi, fr, el, hu, id, hi, fa, `is`, he, it, ja, ko, lt, lv, nl,
     pl, ru, uk, be, es, sl, sv, tr, vi
     
-    var resourceName: String {
+    func resourceName(for fileType: FileType) -> String {
+        var resourceId = ""
+        
         switch self {
```
Comment (Brandon-T) [inline Client/WebFilters/AdblockResourcesMappings.swift:27]: Do we prefer:
```
let map = [.ar: "...", .bg: "...", .zh: "..."]
resourceId = map[fileTypye] ?? ""
```
or `switch` statement instead?

It's just a question. You don't have to change your stuff.

File: Client/WebFilters/ShieldStats/Adblock/AdblockRustEngine.swift [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    
+    func shouldBlock(requestUrl: String, requestHost: String, sourceHost: String) -> Bool {
+        var explicitCancel = false
+        var savedFromException = false
+        let thirdPartyRequest = requestHost != sourceHost
+        
+        var emptyPointer: UnsafeMutablePointer<Int8>?
+        let redirect = UnsafeMutablePointer(mutating: &emptyPointer)
+        
+        return engine_match(engine, requestUrl, requestHost,
```
Comment (Brandon-T) [inline Client/WebFilters/ShieldStats/Adblock/AdblockRustEngine.swift:21]: The `redirect` pointer is passed in but the out value is never used?

Also, do you know if:

`engine_match(...., UnsafeMutablePointer(mutating: &emptyPointer))`

would would instead of creating `redirect`?

File: Client/WebFilters/AdblockResourcesMappings.swift (hunk 3 above)
Comment (iccub) [reply Client/WebFilters/AdblockResourcesMappings.swift:27]: I don't see how this code would go then, `map` is a `locale:id`  dictionary , while `fileType` is a different enum. So I can't reallty do it like that

File: Client/WebFilters/ShieldStats/Adblock/AdblockRustEngine.swift (hunk 4 above)
Comment (iccub) [reply Client/WebFilters/ShieldStats/Adblock/AdblockRustEngine.swift:21]: yes, we don't use any of these inout values but the library requires them, that's why I put this code in a Swift wrapper

I don't understand this part, could you rephrase
```
engine_match(...., UnsafeMutablePointer(mutating: &emptyPointer))
would would instead of creating redirect?
```

I tried passing nil value there but the lib was crashing

In general, this library is used for ad blocking on desktop and Android, while we on iOS use it only for checking if a resource is blocked, that's why most of it is not utilized

Comment (Brandon-T) [reply Client/WebFilters/ShieldStats/Adblock/AdblockRustEngine.swift:21]: @iccub 

```swift
func engine_match(..., ..., redirect: UnsafeMutablePointer<UnsafeMutablePointer<Int8>?>)
```

Instead of doing:
```swift
var emptyPointer: UnsafeMutablePointer<Int8>?
let redirect = UnsafeMutablePointer(mutating: &emptyPointer)

engine_match(..., ..., redirect)
```

Would this work:

```swift
var emptyPointer: UnsafeMutablePointer<Int8>?
engine_match(..., ..., UnsafeMutablePointer(mutating: &emptyPointer))
```

Essentially getting rid of `redirect` as a temporary variable. Just curious if it would compile and work.

Otherwise just leave it how it is :)

**Review by Brandon-T — APPROVED** [review]

File: Client/WebFilters/ShieldStats/Adblock/AdblockRustEngine.swift (hunk 4 above)
Comment (iccub) [reply Client/WebFilters/ShieldStats/Adblock/AdblockRustEngine.swift:21]: it works and compiles, changing it to your version :)

File: Client/WebFilters/AdblockerType.swift [hunk 5]
```diff
@@ -7,7 +7,17 @@ import Shared
...
 
-enum FileType: String { case dat, json, tgz }
+enum FileType: String {
+    case dat, json, tgz
+    
+    var endpoint: String {
+        switch self {
+        case .dat: return "https://adblock-data.s3.brave.com/4/"
+        case .json: return "https://adblock-data.s3.brave.com/ios/"
+        case .tgz: return "" // Httpse resources are not supported yet.
```
Comment (jhreis) [inline Client/WebFilters/AdblockerType.swift:17]: Clarified this, but wanted to say that this does _not_ impact HTTPSE stats, as we currently use a different system for this anyway.

File: Client/WebFilters/AdblockResourcesMappings.swift [hunk 6]
```diff
@@ -14,37 +21,45 @@ enum ResourceLocale: String {
...
+        case .pl: resourceId = "BF9234EB-4CB7-4CED-9FCB-F1FD31B0666C"
+        case .ru, .uk, .be: resourceId = "80470EEC-970F-4F2C-BF6B-4810520C72E6"
+        case .es: resourceId = "AE657374-1851-4DC4-892B-9212B13B15A7"
+        case .sl: resourceId = "418D293D-72A8-4A28-8718-A1EE40A45AAF"
+        case .sv: resourceId = "7DC2AC80-5BBC-49B8-B473-A31A1145CAC1"
+        case .tr: resourceId = "1BE19EFD-9191-4560-878E-30ECA72B5B3C"
+        case .vi: resourceId = "6A0209AC-9869-4FD6-A9DF-039B4200D52C"
+        }
+        
+        switch fileType {
```
Comment (jhreis) [inline Client/WebFilters/AdblockResourcesMappings.swift:59]: Creative solution to get around this 😄

**Review by jhreis — APPROVED** [review]

**Review by jumde — APPROVED** [review]

**Comment by jumde** [comment]
lgtm
